### ISSUE TITLE
論文のTest RMSEまで届かない

### ISSUE THREAD

**Issue by tanimutomo** [tanimutomo/gcmc#15]
論文では，MovieLens 100KのFeatureなしで，0.910のRMSEが出たと書いてあった．
一方で，現時点でテストのRMSEが0.980くらいしか出ていないので，何かが間違っている．
ハイパラとかはだいたい論文内に記載されているのでそれをそのまま使用している．

引っかかっているのはtest時のema decay factorで，これはTraining時のWeight decayとは違くて，学習されたモデルのパラメータの指数移動平均をとっている的なことが書かれていたのだが，よくわからない．
ちゃんと調べる必要がある．
とりあえず，weightの正則化をするweight decayは0に設定して実験中

**Comment by tanimutomo** [comment]
NormalizationはLeftであっているが，
Accumulation functionが'add'ではなく，'stack'であることがわかったので，そこを修正して実験回す．

**Comment by tanimutomo** [comment]
あと違っているのは，
dense layerの

**Comment by tanimutomo** [comment]
pytorch geometricのmessage passingにaggr='stack'を指定できない．．．
MessagePassing classから自分で実装するしかないか．．．

**Comment by tanimutomo** [comment]
ちなみに，lr=0.02のweight_decay=0はいくくて，1.065くらいにとどまった（Test RMSE）
ここからはちゃんとcometml使う

**Comment by tanimutomo** [comment]
stackで，全部のbnとreluをonにすると，lossも両方のrmseも全然下がらない．高い値で止まってしまう．
一方で論文通りrgc layerのreluのみonであとはすベてoffにすると，最初と同じで全く学習できない問題が起こる．

**Comment by tanimutomo** [comment]
もしかしたら1の方法じゃなくて1の方法で，75 / 5がそれぞれのRの特張良になるのかな．．．

**Comment by tanimutomo** [comment]
著者実装をみると，どうやらweight initializationの仕方が違うっぽい
`def weight_variable_random_uniform(input_dim, output_dim=None, name=""):
    """Create a weight variable with variables drawn from a
    random uniform distribution. Parameters used are taken from paper by
    Xavier Glorot and Yoshua Bengio:
    http://jmlr.org/proceedings/papers/v9/glorot10a/glorot10a.pdf"""
    if output_dim is not None:
        init_range = np.sqrt(6.0 / (input_dim + output_dim))
        initial = tf.random_uniform([input_dim, output_dim], minval=-init_range, maxval=init_range, dtype=tf.float32)
    else:
        init_range = np.sqrt(6.0 / input_dim)
        initial = tf.random_uniform([input_dim], minval=-init_range, maxval=init_range, dtype=tf.float32)
    return tf.Variable(initial, name=name)`

**Comment by tanimutomo** [comment]
nn.Linearのところにreset_paramを入れていなかったので，そこらへんを追加すうる

**Comment by tanimutomo** [comment]
weight initializationに関して，reluを入れない場合は6.0じゃなくて2.0とかにしたほうがいいのかな

**Comment by tanimutomo** [comment]
著者実装は基本的にsparse matrixとして扱っているので，u_inputとかにはvalueが入っていて，u_indicesとかにそれらの場所を示すindexが保管してある．おそらくtf.gatherで

**Comment by tanimutomo** [comment]
やはり，dense layerのところでもdrop outして bilinear decoderでもdropoutしている
とりあえずそれを追加して実験

**Comment by tanimutomo** [comment]
あ，もしかしたらtrainから出力された結果のrating matrixを使ってそのままテストしているからいけないのか？
そうするとdropoutとかも自動で適用されてしまってよくわからないことになる説？

**Comment by tanimutomo** [comment]
lossとtrain_rmseが一定期間止まって，少し下がってをloopしている．
test_rmseに関しては1.08あたりで停滞している．
500epoch過ぎてもこの状況なのでこれ以上待っても変化はないかもしれない

**Comment by tanimutomo** [comment]
あと，気になっているのがbilinear decoderに関して，q-matrixのbasis matricesがdiagonalかorthogonal担っていて，明らかにそこのパラメータ数が小さくなっているので，それを実装する必要がある？
もしかしたら，user itemのそれぞれのノードごとの全て計算することを仮定しているので，論文内で勝手に行列として解釈していたものは，全てnodeごとのベクトルに対して適用するように書かれていたのかもしれない．
だとすると，色々問題が出てくるので，一旦論文を読み直して，その可能性がありそうな場所をピックアップする

**Comment by tanimutomo** [comment]
基本的にdiagonalはFalseになっていてTrueになっている場合は存在しないので，おそらく基本的にはQのbasis matrix(P)は，Orthogonal matrixとして定義されている．

**Comment by tanimutomo** [comment]
とりあえず，nn.init.orthogonalで，basis_matrixを直交行列で正則化した

**Comment by tanimutomo** [comment]
ただ，bilineardecoderでrelationごとにuser, itemのfeatureを利用していないので，stackGCN，もしくはDense layerの中でおそらくそれらのデータが集約されているはずなので，それがどこで行われているのかを調べる必要がある．

**Comment by tanimutomo** [comment]
bilinear decoderのbasis-matrixの初期値をorthogonalで設定しても精度の改善は特に見られなかった．

**Comment by tanimutomo** [comment]
おそらくstackの場合でもrelationごとの情報が，Bilinear Decoderでのマトリックス計算の前どこかで集約されている気がする

**Comment by tanimutomo** [comment]
多分わかった
今はstackの場合に新しくrealtionのaxisを増やして計算しているが，実際はその必要性はなくhidden_dimの500や75が5分割されてそれぞれのrelationの特徴量に対応するようになっている．
まさかとは思ったけどそのために5の倍数になっていたのかー．．．
となると，学習がうまくいってなかったのはparameterが多すぎたせいということになるのか
とりあえず，5分割にしてそれぞれにrelaitonを対応させるバージョンの実装を行う．
これで無理だったら諦めてwith feauturesとかの実装をする

**Comment by tanimutomo** [comment]
messageをpropagateする段階でmessageをtarget nodeに代入しているが，そうすると同じtarget nodeに同じrelationできた情報がsumされない問題が起こっていると考えられるので，まず同じtarget nodeに同じrelationでpropagateされるmessageをsumする必要がある

**Comment by tanimutomo** [comment]
結構クソなことに気づいたのが，movielens 10mで使っているsum accumulation function以外は，そもそもordinal weight sharingが実装されていない．
stackGCNの場合は，xavier uniformでinitializationしたweightを100ごとにsplitして，それを各relationの重みとして定義している
そして，最終的にgcn layerの最後にstackして終了している．
論文に書いてないことありすぎてきもい

**Comment by tanimutomo** [comment]
各relationで行うtorch scatter addのところに，全てのEdgesのmessageを入れると実際には与えられるはずのないメッセージがpropagateされてしまう．
というのも，edgeはedge_typeをもっており，該当するtypeのinput * weightはpropagateされて構わないが，他のtypeのmessageはpropagateされるのはおかしいので，そこをedge_typeで判別する必要がある気がする．
つまり，全てのedgeのmessageから該当するtypeだけに絞ってtorch scatter addをしなければならない．
基本的には，messageを送る先のnew embedding matrixはtorch_scatterが勝手に形成をしてくれるので，ちゃんとmessageを絞るのと，それ該当するindexも絞って遅れば大丈夫な気がする．

**Comment by tanimutomo** [comment]
一応，RGClayerから出力される中間層におけるdimentionはあらかじめ定義されていて(500)，それを各relationの特徴量をsumして実現するか，100ずつのrelationの特徴量としてstackして実現するかという違いにしているらしい．非常に分かりにくいが実装を見るとそうなっていることが伺える

**Comment by tanimutomo** [comment]
LossもRMSEも1.14くらいと，全然下がっていない．
ただ1000epoch目でもLossが下がっていて，RMSEも微笑ではあるが下がっていることから，もしかすると収束が遅いだけかもしれないので，BNを追加して実験中

**Comment by tanimutomo** [comment]
変更するべき箇所としては，splitしてaddするときにrelationで絞る必要性がないかもしれない（理論上はやるべきだと思うが著者実装ではそのようなことはされていない）
また，scatter_addの初期値が気になる．そこで該当するものがない場合に変な値で埋められていないかということ．（でもdocumentを見る限り，addされない箇所は0になっているっぽいので，心配ないかも．）

**Comment by tanimutomo** [comment]
明らかにLossが高止まりしている．．．

**Comment by tanimutomo** [comment]
TestRMSEは明らかに下がっているが，そもそもがかなり高いしなだらかに低下しているので，0.910までは届きそうにない気がする

**Comment by tanimutomo** [comment]
100ごとの特徴量を抽出してきて，それらをconcatするときにrelationごとに絞ってそれらのsumをとってからconcatしないと，最終的にただただaddをしているのと同じになってしまうので，精度改善されたと思ったけど，実験として意味をなしていない

**Comment by tanimutomo** [comment]
decoderのdropoutはoffにしているっぽい

**Comment by tanimutomo** [comment]
論文内では入力行列のXに，Content featureを入れてしまうと，この場合はContent featureがそれぞれのノードを識別できるだけの情報量を持っていないために，情報のフローでボトルネックが発生してしまう．
そのため，ここではXは各ノードを識別するためのIdentity matrixとすると言っていたが，実装を見るとidentity matrixの後にuserとitemの次元数を合わせるためのぜろ行列をconcatしている．
これでuserとitemのweight sharingをrgcn layerでも実現していると思われる．
そうなると，ちゃんとxとwをmultiplyしないといけない可能性が出てくるのか
そして，uとiの特徴量を分けて計算しないといけなくなる可能性が高いな

**Comment by tanimutomo** [comment]
そして，指数移動平均の話をすっ飛ばしていることに気づいた．
でもPytorchにはその機能はなさそうなのと，実装する時間はなさそうなので，やめておく

**Comment by tanimutomo** [comment]
そもそも最初からuserとitemの特徴行列をXとしてまとめないで，最初から分けて入力する

**Comment by tanimutomo** [comment]
userとitemのweight shareはrgc layerでもdenseでもやってる
また入力行列をuserとitemに分けたりそれらをweightとmatmulする必要はない．
後方にゼロ行列をconcatしただけなので，weightのを取ってくる範囲をidentity matrixまでの範囲で指定してsliceすれば，matmulと同等のことが計算量ほぼ0で行える

**Comment by tanimutomo** [comment]
node dropoutを自作で実装したけど，あれが間違っている可能性が高い．
dropoutの原理はもっと複雑になっている気がする．．．

**Comment by tanimutomo** [comment]
identity matrixとzero matrixのconcatに対して積をとるだけであれば，実装する手間はそこまででもないので，もし今の実装でうまくいかなかったら試してみてもいいかもしれない．
node-dropoutに関しては，著者実装を真似しているのであっている可能性も高い．
ただただ各ノードのからの情報が伝播しないだけなので，普通のdropoutとは異なるのかもしれない．

**Comment by tanimutomo** [comment]
一旦しめる
